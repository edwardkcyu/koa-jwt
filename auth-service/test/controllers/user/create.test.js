require('dotenv').config({ path: 'test/.env.test' });

const request = require('supertest');

const { User } = require('../../../src/models');
const app = require('../../../src/app');
const { HTTP_STATUS } = require('../../../src/lib/constants');

describe('POST /users', () => {
  const url = '/users';

  beforeEach(() => {
    User.findOne = jest.fn(() => null);

    User.prototype.save = jest.fn(() => ({
      _id: '1a2b3c',
      userName: 'curry',
      hashedPassword: 'cba0021ca36477ed4be95626316ac66e31ef7d73e8f9e3888e36737e1a9ae597'
    }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  afterAll(() => {
    app.server.shutdown(function() {
      console.log('Everything is cleanly shutdown.');
    });
  });

  test('successfully registers a user', async () => {
    const response = await request(app.callback())
      .post(url)
      .send({ userName: 'curry', password: '123' })
      .set('Content-Type', 'application/json');

    expect(response.status).toEqual(HTTP_STATUS.CREATED);
    expect(response.type).toEqual('application/json');
    expect(response.body).toMatchSnapshot({
      createdAt: expect.any(String)
    });
  });

  test('responses error if user name is empty', async () => {
    const response = await request(app.callback())
      .post(url)
      .send({ password: '123' })
      .set('Content-Type', 'application/json');

    expect(response.status).toEqual(HTTP_STATUS.BAD_REQUEST);
    expect(response.type).toEqual('application/json');
    expect(response.body).toMatchSnapshot();
  });

  test('responses error if password is empty', async () => {
    const response = await request(app.callback())
      .post(url)
      .send({ userName: 'curry' })
      .set('Content-Type', 'application/json');

    expect(response.status).toEqual(HTTP_STATUS.BAD_REQUEST);
    expect(response.type).toEqual('application/json');
    expect(response.body).toMatchSnapshot();
  });

  test('responses error if user already exists', async () => {
    User.findOne = jest.fn(() => ({
      _id: '1a2b3c',
      userName: 'curry',
      hashedPassword: 'cba0021ca36477ed4be95626316ac66e31ef7d73e8f9e3888e36737e1a9ae597'
    }));

    const response = await request(app.callback())
      .post(url)
      .send({ userName: 'curry', password: '123' })
      .set('Content-Type', 'application/json');

    expect(response.status).toEqual(HTTP_STATUS.BAD_REQUEST);
    expect(response.type).toEqual('application/json');
    expect(response.body).toMatchSnapshot();
  });
});
