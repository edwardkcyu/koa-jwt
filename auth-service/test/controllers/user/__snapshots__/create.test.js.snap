// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /users responses error if password is empty 1`] = `
Object {
  "error": Object {
    "code": "E002",
    "message": "Password should not be empty",
  },
}
`;

exports[`POST /users responses error if user already exists 1`] = `
Object {
  "error": Object {
    "code": "E003",
    "message": "User(curry) already exists",
  },
}
`;

exports[`POST /users responses error if user name is empty 1`] = `
Object {
  "error": Object {
    "code": "E001",
    "message": "User name should not be empty",
  },
}
`;

exports[`POST /users successfully registers a user 1`] = `
Object {
  "createdAt": Any<String>,
  "userId": "1a2b3c",
  "userName": "curry",
}
`;
